### PR TITLE
Add color pass flags to Forward Clustered renderer

### PR DESCRIPTION
**Pull request by JFonS**
This PR removes a lot of enum values related to the color render pass in favor of a new flag-bases approach. This means instead of hard-coding all the possible option combinations into enums, we can write our logic by checking a bit-mask.

The changes in `rendering_device_vulkan.cpp` add support for unused attachments. That means `RenderingDeviceVulkan::framebuffer_create()` can take null `RID`s in the attachments vector, which will result in `VK_ATTACHMENT_UNUSED` entries in the render pass.

This is used in this same PR to establish fixed locations for the color pass attachments (only color and separate specular so far, but TAA will add motion vectors as well). This way, the attachment locations in the shader can stay the same regardless of which attachments are actually used.

Right now all the combinations of flags are generated, but we will need to add a way to limit the amount of combinations in the future.

I still need to look into whether this makes sense in the mobile renderer, but I wanted to get the PR out to get some feedback.

@BastiaanOlij I had to rebase on top of your multiview PR. I'm pretty sure my changes result in the same pipelines and shader versions as the ones you added, but I'm not sure how to properly test multiview, so if you can give it a check that would be awesome.

### CODE REVIEW

**Comment by BastiaanOlij** [comment]
This looks awesome! We should also apply these changes to the mobile renderer.

As for stereoscopic, it crashes, strangely enough even in the editor:
```
CrashHandlerException: Program crashed
Engine version: Godot Engine v4.0.alpha.custom_build (60ae040b56b87e4cb5894c84c57abc6a0749368f)
Dumping the backtrace. Please include this when reporting the bug on https://github.com/godotengine/godot/issues
[0] Map<unsigned int,RID,Comparator<unsigned int>,DefaultAllocator>::find (D:\Development\godot4-git\core\templates\map.h:585)
[1] Map<unsigned int,RID,Comparator<unsigned int>,DefaultAllocator>::has (D:\Development\godot4-git\core\templates\map.h:621)
[2] RendererSceneRenderImplementation::RenderForwardClustered::RenderBufferDataForwardClustered::get_color_pass_fb (D:\Development\godot4-git\servers\rendering\renderer_rd\forward_clustered\render_forward_clustered.cpp:243)
[3] RendererSceneRenderImplementation::RenderForwardClustered::_render_scene (D:\Development\godot4-git\servers\rendering\renderer_rd\forward_clustered\render_forward_clustered.cpp:1630)
[4] RendererSceneRenderRD::render_scene (D:\Development\godot4-git\servers\rendering\renderer_rd\renderer_scene_render_rd.cpp:5096)
[5] RendererSceneCull::_render_scene (D:\Development\godot4-git\servers\rendering\renderer_scene_cull.cpp:3223)
[6] RendererSceneCull::_render_reflection_probe_step (D:\Development\godot4-git\servers\rendering\renderer_scene_cull.cpp:3344)
[7] RendererSceneCull::render_probes (D:\Development\godot4-git\servers\rendering\renderer_scene_cull.cpp:3372)
[8] RenderingServerDefault::_draw (D:\Development\godot4-git\servers\rendering\rendering_server_default.cpp:91)
[9] RenderingServerDefault::draw (D:\Development\godot4-git\servers\rendering\rendering_server_default.cpp:383)
[10] Main::iteration (D:\Development\godot4-git\main\main.cpp:2682)
[11] OS_Windows::run (D:\Development\godot4-git\platform\windows\os_windows.cpp:667)
[12] widechar_main (D:\Development\godot4-git\platform\windows\godot_windows.cpp:169)
[13] _main (D:\Development\godot4-git\platform\windows\godot_windows.cpp:191)
[14] main (D:\Development\godot4-git\platform\windows\godot_windows.cpp:205)
[15] WinMain (D:\Development\godot4-git\platform\windows\godot_windows.cpp:219)
[16] __scrt_common_main_seh (d:\a01\_work\6\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288)
[17] BaseThreadInitThunk
-- END OF BACKTRACE --
```

Here is a simple test project that works without the need of XR equipment as it just outputs a side by side image:
https://github.com/BastiaanOlij/godot_test_multiview

**Comment by JFonS** [comment]
@BastiaanOlij the crash was related to reflection probes, but your test project helped me fix that bug and other multiview related bugs. I force-pushed the updated version.

In your project, I'm still getting some weird issues that seem related to sky rendering, and I can reproduce them on current `master` too, so I think they are not related to my changes, but it's hard to tell if my PR breaks anything else until we are sure it works fine on master.

**Comment by JFonS** [comment]
@BastiaanOlij I pushed a new version, everything should be working now.

**Comment by JFonS** [comment]
@reduz I pushed a new version that allows for registering the valid flag combinations and only compiles those pipelines. The PR is ready for a final review and merge.

**Comment by BastiaanOlij** [comment]
> @BastiaanOlij I pushed a new version, everything should be working now.

Cool, what turned out to be the problem?

**Comment by JFonS** [comment]
@BastiaanOlij I had accidentally removed a fix while getting rid of some debug prints :)

**Review by BastiaanOlij — APPROVED** [review]
This looks good to me, should probably be applied to the mobile renderer as well

File: drivers/vulkan/rendering_device_vulkan.cpp [hunk 1]
```diff
@@ -3364,8 +3364,14 @@ VkRenderPass RenderingDeviceVulkan::_render_pass_create(const Vector<AttachmentF
 	attachment_last_pass.resize(p_attachments.size());
 
 	Vector<VkAttachmentDescription> attachments;
+	Vector<int> attachment_remap;
 
 	for (int i = 0; i < p_attachments.size(); i++) {
+		if (p_attachments[i].usage_flags == 0xFFFFFFFF) {
```
Comment (reduz) [inline drivers/vulkan/rendering_device_vulkan.cpp:3370]: I would add a enum { UNUSED_ATTACHMENT = 0xFFFFFFFF }; to RenderingDevice::AttachmentFormat

**Review by reduz — APPROVED** [review]

**Comment by reduz** [comment]
Seems good, pending fix on above comment to merge.

**Comment by JFonS** [comment]
Rebased and updated, should be good to merge.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**